### PR TITLE
development_workflow: expand 'squash' and 'rebase' developer documentation

### PR DESCRIPTION
**Pull request by sladen**
Expand *squash* and *rebase* developer documentation based on discussion from end of [bug #5869].

Kudos to @pllim, @bsipocz, @astrofrog, @MSeifert04 for contributing to the discussion.

_(Hopefully this accurately reflects the discussion, feedback and suggestions for further refinement welcomed)._

### CODE REVIEW

File: docs/development/workflow/development_workflow.rst [hunk 1]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
 Rebase, but only if asked
 =========================
 
-Sometimes the maintainers of Astropy will ask you to *rebase* your changes
-before they are merged into the main Astropy repository.
+Sometimes the maintainers of Astropy may ask a pull request to be *rebased*
+(`git rebase`) or *squashed*/*squished* (`git merge --squash`) before
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: really `squish`?

File: docs/development/workflow/development_workflow.rst [hunk 2]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
+changes in a Pull Request are merged into the main Astropy
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: it's not only "before" merged, sometimes a rebase is necessary because of other merged commits and CI will refuse to start.

File: docs/development/workflow/development_workflow.rst [hunk 3]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
+*master* repository.
+
+The decisions of when to request a *squash* or *rebase* are left to
+individual maintainers.  These may be requested done to reduce the number of
+visible commits saved in the repositary history, or becauses of code-changes
+in Astropy in the mean-time.  Both involve rewriting the Git history, meaning
+that commit IDs will change, which is why you should do it only if asked.
```
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: "IDs" -> "hash" (also apply to all the others below)

File: docs/development/workflow/development_workflow.rst [hunk 4]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
 
 It is easier to make mistakes rebasing than other areas of `git`_, so before you
 start make a branch to serve as a backup copy of your work::
 
     git branch tmp my-new-feature # make temporary branch--will be deleted later
 
+After altering the history with `git rebase` or `git merge --squash` a normal
+`git push` is prevented, and for both a `git push --force` will be required.
+
+... _howto_rebase:
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: should be only two `..`.

File: docs/development/workflow/development_workflow.rst [hunk 5]
```diff
@@ -460,17 +478,49 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+===========
+
+After using `git rebase` or `git merge --squash` you will still need
+to push your changes to GitHub so that they are visible to others and
+the Pull Request can be updated.  Use of a simple `git push` will be
+prevented because of the changed history, and will need to be manually
+overridden using::
 
-    git push -f
+    git push --force
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: well, these are equivalent, any reason to change this?

File: docs/development/workflow/development_workflow.rst [hunk 6]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
 
+After altering the history with `git rebase` or `git merge --squash` a normal
+`git push` is prevented, and for both a `git push --force` will be required.
+
+... _howto_rebase:
+
+How to rebase
+=============
+
+Behind the scenes, `git`_ is deleting the changes and branch you made, making the
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: I think this is a bit too much information for this section, the mechanics for `git` are explained in a lot of tutorials. I think we should focus on the "how" here.

File: docs/development/workflow/development_workflow.rst [hunk 7]
```diff
@@ -460,17 +478,49 @@ You are more likely to run into *conflicts* here--places where the changes you
 made conflict with changes that someone else made--than anywhere else. Ask for
 help if you need it.
 
-After the rebase you need to push your changes to GitHub; you will need force
-the push because `git`_ objects to re-writing the history of the repository
-after you have pushed it somewhere::
+... _howto_squash:
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: only two `..`

File: docs/development/workflow/development_workflow.rst (hunk 1 above)
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: 1. Use double backticks (also apply to all other single backticks on command line examples
```
``git rebase``
```

2. I have never seen anyone write "squish the commits" but I guess it doesn't hurt...

File: docs/development/workflow/development_workflow.rst [hunk 8]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
+metadata/commit IDs, and also removes separate visibility of individual commits;
+a new commit and commit message will only contain a textual list of the earlier
+commits.
 
 It is easier to make mistakes rebasing than other areas of `git`_, so before you
 start make a branch to serve as a backup copy of your work::
 
     git branch tmp my-new-feature # make temporary branch--will be deleted later
 
+After altering the history with `git rebase` or `git merge --squash` a normal
```
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: Add comma before "a normal".

File: docs/development/workflow/development_workflow.rst [hunk 9]
```diff
@@ -460,17 +478,49 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+Typically we ask to *squash*/*squish* when there was a fair amount of trial
+and error, but the final patch remains quite small, or when files were added
+and removed (especially binary files or files that should not remain in the
+repository) or if the number of commits in the history is disproportionate
+compared to the work being carried out (for example 30 commits gradually
+refining a final 10-line change).  Conceptually this is equivalent to
+exporting the final diff from a feature branch, then starting a new branch and
+only applying only that patch:
+
+The actual squashing is done on a new branch, using `git merge --squash`::
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: I always used `git rebase -i HEAD~n`. Avoids the whole copy-branch stuff. Not sure which one is "easier" or "better" though.

File: docs/development/workflow/development_workflow.rst [hunk 10]
```diff
@@ -460,17 +478,49 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+=============
+
+Typically we ask to *squash*/*squish* when there was a fair amount of trial
+and error, but the final patch remains quite small, or when files were added
+and removed (especially binary files or files that should not remain in the
+repository) or if the number of commits in the history is disproportionate
+compared to the work being carried out (for example 30 commits gradually
+refining a final 10-line change).  Conceptually this is equivalent to
+exporting the final diff from a feature branch, then starting a new branch and
+only applying only that patch:
```
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: Replace ":" with ".".

File: docs/development/workflow/development_workflow.rst [hunk 11]
```diff
@@ -460,17 +478,49 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+
+The actual squashing is done on a new branch, using `git merge --squash`::
+
+  git checkout master -b squashed-new-feature
```
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: This is not how I usually squash commits. I remain in the PR branch and then I do:
```
git rebase -i HEAD~n
```
where `n` is the number of commits to squash, then you follow the prompted instructions.

File: docs/development/workflow/development_workflow.rst (hunk 5 above)
Comment (pllim) [inline docs/development/workflow/development_workflow.rst]: Why change `-f` to `--force`? The former works too and is actually less typing.

Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @MSeifert04:  Slight verbosity, so that it's clearly to the user what they are doing ("a force", not "a file" etc), and that it is intended to be exceptional usage, not normal practice.

Happy to revert it if that's easier.

File: docs/development/workflow/development_workflow.rst (hunk 2 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: maybe rephrase along the lines of "when there is a conflict with the current master", but you already say that in the next paragraph.

File: docs/development/workflow/development_workflow.rst (hunk 5 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: I very much 👍  for this change, being explicit is never bad in a tutorial.

**Comment by pllim** [comment]
Ops... @MSeifert04 and I are at it again with concurrent reviews! I apologize for duplicate comments.

File: docs/development/workflow/development_workflow.rst (hunk 1 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @MSeifert04: Added to improve searchability, will nuke it, since it was _me_ that used it…

File: docs/development/workflow/development_workflow.rst (hunk 9 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: I also always do the squash as part of a rebase (when it can be a rebase to the base of the PR as @MSeifert04 says). 
Also, I think a user should never ever use the ``git merge`` command as it's more likely to cause more mess than be used properly.

File: docs/development/workflow/development_workflow.rst (hunk 1 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: Please use double backticks for command quotes, otherwise sphinx will complain for the missing API docs.

File: docs/development/workflow/development_workflow.rst (hunk 2 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: CI note added.

File: docs/development/workflow/development_workflow.rst (hunk 6 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @MSeifert04: the "Behind the scenes" part is from the existing documentation, happy to adjust as required.

File: docs/development/workflow/development_workflow.rst (hunk 4 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: one

File: docs/development/workflow/development_workflow.rst (hunk 7 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done

File: docs/development/workflow/development_workflow.rst (hunk 5 above)
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: My preference is `-f` but if you like the verbosity I'm not opposed to `--force`.

File: docs/development/workflow/development_workflow.rst (hunk 9 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @MSeifert04, and and copying the branches isn't required, and there are various other methods.  Using the built-in `git merge --squash` is hopefully more approachable for the first time easier, and moving a branch with `-m .. tmp` instead of `-M squash-new` leaves the state consistent for the combined `git push --force && git branch -D tmp` documentation at the end

File: docs/development/workflow/development_workflow.rst (hunk 4 above)
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: no, two :)

File: docs/development/workflow/development_workflow.rst (hunk 6 above)
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: in that case maybe someone else should decide if we should keep it. I just felt it could be distracting but someone thought different. @bsipocz ?

File: docs/development/workflow/development_workflow.rst (hunk 4 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: d-d-d…doh!

File: docs/development/workflow/development_workflow.rst (hunk 1 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @pllim: double backticks done (Restructured Text vs. Github).  Removed squish.

File: docs/development/workflow/development_workflow.rst (hunk 2 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: (1) done (down-cased), (2) already reworded.

File: docs/development/workflow/development_workflow.rst [hunk 12]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
 
-Sometimes the maintainers of Astropy will ask you to *rebase* your changes
-before they are merged into the main Astropy repository.
+Sometimes the maintainers of Astropy may ask a pull request to be *rebased*
+(`git rebase`) or *squashed*/*squished* (`git merge --squash`) before
+changes in a Pull Request are merged into the main Astropy
+*master* repository.
+
+The decisions of when to request a *squash* or *rebase* are left to
+individual maintainers.  These may be requested done to reduce the number of
```
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done!

File: docs/development/workflow/development_workflow.rst [hunk 13]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
+visible commits saved in the repositary history, or becauses of code-changes
```
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: Hopefully all done

File: docs/development/workflow/development_workflow.rst [hunk 14]
```diff
@@ -433,24 +433,42 @@ to GitHub. GitHub will automatically update your pull request.
...
+in Astropy in the mean-time.  Both involve rewriting the Git history, meaning
```
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done

File: docs/development/workflow/development_workflow.rst (hunk 3 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done (hashes, hashes, hash).

File: docs/development/workflow/development_workflow.rst (hunk 8 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done

File: docs/development/workflow/development_workflow.rst (hunk 10 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done

File: docs/development/workflow/development_workflow.rst (hunk 5 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: Happy either way.  _(See discussion by @MSeifert04 and @bsipocz above)._

File: docs/development/workflow/development_workflow.rst (hunk 1 above)
Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: done!

**Comment by sladen** [comment]
Wow.

Thank you for all for the feedback coming so fast.

Commit f8bdaab diff contains a a possible _Squash using rebase_ section; depending on people's preferences about how recoverable/undoable the example method should be.

The other two places that need further discernment appear to be

1. `git push -f` vs `git push --force` abbreviation (or not)
2. Keeping/altering existing "Behind the scenes" paragraph introducing for `git rebase`.

File: docs/development/workflow/development_workflow.rst (hunk 5 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: This guide is written for people new to git, so we should write out ``--force`` to emphasize that this should not be done by default. Obviously when someone is better acquaint with git (e.g *reads* it manpage) will discover that ``-f`` does the same job with less typing.

File: docs/development/workflow/development_workflow.rst (hunk 9 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: I still feel that his way is weird, and more prone to mess-ups. Also what you do is basically a rebase as the new branch starts from master (btw I'm in the middle of another edit to make sure we advise people to delete their master, so better not to refer to it elsewhere in the docs, all new branches should be branched out of the remote astropy/master).

So I'm -1 on adding this section, ``git --rebase -i`` is a swiss army knife, and should be enough to show to new people here. They can still read up on other gitfu on git tutorial pages. 
However I'm asking @eteq, @astrofrog, @mhvk, @cdeil, @pllim and @Cadair about their opinion as they are the ones running into new people and teaching these steps.

File: docs/development/workflow/development_workflow.rst [hunk 15]
```diff
@@ -460,17 +479,62 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+Typically we ask to *squash* when there was a fair amount of trial
+and error, but the final patch remains quite small, or when files were added
+and removed (especially binary files or files that should not remain in the
+repository) or if the number of commits in the history is disproportionate
+compared to the work being carried out (for example 30 commits gradually
+refining a final 10-line change).  Conceptually this is equivalent to
+exporting the final diff from a feature branch, then starting a new branch and
+applying only that patch.
+
+The actual squashing is done on a new branch, using ``git merge --squash``::
```
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: I still feel that his way is weird, and more prone to mess-ups. Also what you do is basically a rebase as the new branch starts from master (btw I'm in the middle of another edit to make sure we advise people to delete their master, so better not to refer to it elsewhere in the docs, all new branches should be branched out of the remote astropy/master).

So I'm -1 on adding this section, git --rebase -i is a swiss army knife, and should be enough to show to new people here. They can still read up on other gitfu on git tutorial pages.
However I'm asking @eteq, @astrofrog, @mhvk, @cdeil, @pllim and @Cadair about their opinion as they are the ones running into new people and teaching these steps.

Comment (mhvk) [inline docs/development/workflow/development_workflow.rst]: @bsipocz - the only thing I've ever used is `git rebase -i`, and it is what I always recommend. I think it is particularly useful since it at the same time allows one to rebase on upstream master, i.e., the workflow I like to suggest is:
```
git fetch upstream
git rebase -i upstream/master
git push --force
```

Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: So, what list of commands (precisely) would people prefer to have in its place for new users to be shown?

Comment (mhvk) [inline docs/development/workflow/development_workflow.rst]: The three I showed above...

Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: @mhvk - We could even leave out the fetch upstream, though I also think it's good practice (especially from the CI point of view) and clearly simpler than figuring out the the HEAD numbers

Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: @MSeifert04 - If there are conflicts then github shows it straight away, so one needs to do a rebase to master. If there are no conflicts spotted by github, it's almost certain that the rebase will be painless, too.

Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @mhvk: pushed.

_(It's probably quickly if others could push any ideas/improves to the PR branch directly).

File: docs/development/workflow/development_workflow.rst [hunk 16]
```diff
@@ -460,17 +479,46 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+refining a final 10-line change).  Conceptually this is equivalent to
+exporting the final diff from a feature branch, then starting a new branch and
+applying only that patch.
+
+A squash may also be performed staying within the existing branch using::
+
+  git fetch upstream
+  git rebase -i upstream/master
+
+Where ``n`` is the number of commits to per squashed.  This method
```
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: this is superflous now.

File: docs/development/workflow/development_workflow.rst [hunk 17]
```diff
@@ -460,17 +479,46 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+=============
+
+Typically we ask to *squash* when there was a fair amount of trial
+and error, but the final patch remains quite small, or when files were added
+and removed (especially binary files or files that should not remain in the
+repository) or if the number of commits in the history is disproportionate
+compared to the work being carried out (for example 30 commits gradually
+refining a final 10-line change).  Conceptually this is equivalent to
+exporting the final diff from a feature branch, then starting a new branch and
+applying only that patch.
```
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: I really like to have this sentence for explanation of what happens.

Comment (sladen) [inline docs/development/workflow/development_workflow.rst]: @bsipocz, done.  Thanks for spotting.

File: docs/development/workflow/development_workflow.rst (hunk 6 above)
Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: Maybe a bit of too much info, but I think we can keep it. Keeping purely the ``how`` only will make it even more black box and just adds to the widespread hate for git (though I think it's awesome, I could hardly convince people around me).

**Comment by mhvk** [comment]
@sladen - per your request, I added a bit more explanation directly to the PR. Overall, this looks good and helpful!

File: docs/development/workflow/development_workflow.rst [hunk 18]
```diff
@@ -460,17 +479,48 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+editing has the instructions on what to do.
+
+.. _howto_push_force:
+
+How to push
+===========
+
+After using ``git rebase`` or ``git merge --squash`` you will still need
+to push your changes to GitHub so that they are visible to others and
+the pull request can be updated.  Use of a simple `git push` will be
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: You also need double backticks around this `git push` so that travis passes :)

File: docs/development/workflow/development_workflow.rst [hunk 19]
```diff
@@ -433,24 +433,43 @@ to GitHub. GitHub will automatically update your pull request.
...
+metadata/commit hash, and also removes separate visibility of individual
+commits; a new commit and commit message will only contain a textual
+list of the earlier commits.
 
 It is easier to make mistakes rebasing than other areas of `git`_, so before you
 start make a branch to serve as a backup copy of your work::
 
     git branch tmp my-new-feature # make temporary branch--will be deleted later
 
+After altering the history with ``git rebase`` or ``git merge --squash``,
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: I think one could omit the `git merge` here because we changed it to `rebase`.

File: docs/development/workflow/development_workflow.rst [hunk 20]
```diff
@@ -460,17 +479,48 @@ You are more likely to run into *conflicts* here--places where the changes you
...
+The last command will open an editor with all your commits, allowing you to
+squash several commits together, rename them, etc. Helpfully, the file you are
+editing has the instructions on what to do.
+
+.. _howto_push_force:
+
+How to push
+===========
+
+After using ``git rebase`` or ``git merge --squash`` you will still need
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: again we probably shouldn't mention `git merge` now that we changed it to `rebase -i`

**Comment by MSeifert04** [comment]
CI failure seems unrelated. Otherwise LGTM modulo two small comments.

File: docs/development/workflow/development_workflow.rst [hunk 21]
```diff
@@ -225,7 +225,7 @@ At this point you have made and checked out a new branch, but `git`_ does not
 know it should be connected to your fork on GitHub. You need that connection
 for your proposed changes to be managed by the Astropy maintainers on GitHub.
 
-To connect your local branch to GitHub, you `git push`_ this new branch up to
+To connect your local branch to GitHub, you ``git push`` this new branch up to
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: Why this change? The link works in the current version.

**Comment by bsipocz** [comment]
@MSeifert04 - I've pushed a commit to remove all mentions of ``merge squash``. Have a look and let me know if you're happy to get this merged.

File: docs/development/workflow/development_workflow.rst [hunk 22]
```diff
@@ -457,9 +457,8 @@ start make a branch to serve as a backup copy of your work::
 
     git branch tmp my-new-feature # make temporary branch--will be deleted later
 
-After altering the history with ``git rebase`` or ``git merge --squash``,
-a normal ``git push`` is prevented, and for both a ``git push
---force`` will be required.
+After altering the history, e.g. with ``git rebase``, a normal ``git push``
+is prevented, and for both a ``git push --force`` will be required.
```
Comment (MSeifert04) [inline docs/development/workflow/development_workflow.rst]: "both"?

Comment (bsipocz) [inline docs/development/workflow/development_workflow.rst]: haha, you indeed find all inconsistencies. Will fix and push an amend and merge once the docs build passed.

**Comment by bsipocz** [comment]
OK, enough said and it's time to merge this. Thanks @sladen for this improvement.
